### ISSUE TITLE
Memory Leak when tf.Session run on the sliced tensor

### ISSUE THREAD

**Issue by chychen** [tensorflow/tensorflow#19049]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.5.0-0-g37aa430d84 1.5.0
- **Python version**: 3.6
- **CUDA/cuDNN version**: release 9.0, V9.0.176
- **GPU model and memory**: 1070 8G

### Describe the Feature Request
For current version of Tensorflow, it will leak memory if we evaluate the sliced tensor as attached code and images, which isn't intuitive and hard to debug. It seems like the unused part of tensor will leak memory because there is no python object reference on it. I am not sure whether it could fixed or not, maybe it could show warning at least.

### Source code / logs

`import tensorflow as tf`
`from pympler.tracker import SummaryTracker`
`tracker = SummaryTracker()`
`a = tf.zeros([3,4,5])`
`sess = tf.Session()`

`def leak_version():`
`____return sess.run(a[0,0])`

`def safe_version():`
`____return sess.run(a)[0,0]` 

`for i in range(10):`
`____tracker.print_diff()`
`____# b = safe_version()`
`____b = leak_version()`

#### Trace of the Leak
![image](https://user-images.githubusercontent.com/5878561/39568787-ed1af628-4ef5-11e8-80a0-b1e532b8c81c.png)

#### Leak Version : visualize the object reference graph on object 'b' with depth=5 by Library objgraph
![image](https://user-images.githubusercontent.com/5878561/39569067-adc03b18-4ef6-11e8-8d68-b5dee8a89eb0.png)

#### Safe Version : visualize the object reference graph on object 'b' with depth=5 by Library objgraph
![image](https://user-images.githubusercontent.com/5878561/39568928-57ccc000-4ef6-11e8-871d-7507a6dfefb4.png)

**Comment by jxcodetw** [comment]
Tested with https://pypi.org/project/memory_profiler/
It shows that the total memory usage increases after each leak_version() function call
![test](https://user-images.githubusercontent.com/5556203/39577008-ae8cb960-4f12-11e8-9e23-38c3a082a117.png)

**Comment by jart** [comment]
@rmlarsen There may be a memory leak in `tf.strided_slice`.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 22 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 37 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 52 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 67 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 82 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 97 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 112 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 127 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 142 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 157 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 172 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 187 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 202 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by devstein** [comment]
@rmlarsen 

I'm having the same issue in `tf 2.0.0beta1`. Here is the exact code causing it:

```
def map_to_xy_dataset(csv_dataset, params):
    window_size = params[WINDOW_SIZE_KEY]
    window_shift = params[WINDOW_SHIFT_KEY]
    n_workers = tf.data.experimental.AUTOTUNE

    # THESE SLICES CAUS THE MEMORY LEAK
    def split_xy(window):
        # For X, select all but the last value and flatten
        range_x = tf.shape(window)[1] - 1
        x = tf.reshape(window[:, 0:range_x], [-1])

        # Select a single Y value
        y = window[0, -1]

        return x, y

    # Turn the csv dataset row x col tensor
    row_dataset = csv_dataset.map(lambda *x: tf.reshape(x, [len(x)]))

    windowed_dataset = row_dataset.window(
        size=window_size, shift=window_shift,
        drop_remainder=True).flat_map(lambda x: x.batch(window_size))

    xy_dataset = windowed_dataset.map(split_xy, num_parallel_calls=n_workers)

    return xy_dataset
```

**Comment by kyamagu** [comment]
I am observing the same issue in `tensorflow==1.15.0`. I insert `tf.stop_gradient` to prevent from this leak to happen

**Comment by jackd** [comment]
The leaky version adds a new element to the symbolic graph each time you call it. You should be fine if you create it outside the loop.

```python
import tensorflow as tf
from pympler.tracker import SummaryTracker
tracker = SummaryTracker()
a = tf.zeros([3,4,5])
a00 = a[0, 0]
sess = tf.Session()

def leak_version():
    return sess.run(a)

def safe_version():
    return sess.run(a)[0,0]

for i in range(10):
    tracker.print_diff()
    # b = safe_version()
    b = leak_version()
```

I don't believe this is related to the dataset caching issue (the problem that led me here :S)

**Comment by tilakrayal** [comment]
@chychen,
I tried to execute the code with the alternative approach and was not able to find any issues. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/075b2655358ab17e6dee4bc416985b50/untitled483.ipynb) and let us know if the issue still persists. Thank you!

**Comment by chychen** [comment]
uh... it has been 4 years... maybe we could just leave it? thanks.

**Comment by tilakrayal** [comment]
@chychen,
Could you please feel free to move this issue to closed status. Thank you!
